[JSC] Implement CanDeclareGlobalFunction abstract operation and friends

shvaikalesh opened the pull request:
#### 14f1a47bd7264df4cf795d6ddb88f6e321e8701f
<pre>
[JSC] Implement CanDeclareGlobalFunction abstract operation and friends
<a href="https://bugs.webkit.org/show_bug.cgi?id=260487">https://bugs.webkit.org/show_bug.cgi?id=260487</a>
&lt;rdar://problem/114215396&gt;

Reviewed by Yusuke Suzuki.

For both global code and sloppy mode eval() in global scope, this change adds CanDeclareGlobalFunction [1]
that checks if a function can be declared while conforming to invariants of essential internal methods [2],
and also introduces CreateGlobalFunctionBinding [3] that creates a binding with best possible descriptor,
attempting to make the property both writable and enumerable.

For global code, the presence of these helpers is observable when attempting to shadow a non-configurable
global propeperty, and in browser environments with multiple &lt;script&gt; elements: if first script defines
non-configurable property X, the second one containing `function X() {}` should throw a TypeError.
In Bun, they can be observed via vm.runInThisContext() API.

Just like before the change, global code functions are defined on the symbol table unless there is a
non-configurable structure property, ensuring the best possible performance. CanDeclareGlobalFunction
and CreateGlobalFunctionBinding substitute the usage of non-idiomatic VM::DeletePropertyModeScope hack,
which used to ensure there is no property by the same name both on the structure and on the symbol table.
This change makes its removal one step closer.

Functions declared via sloppy mode eval() in global scope are unconditionally defined on the structure
because JSC relies on symbol table entries being immutable. This also preserves a bit weird global object
property enumeration order that is different from both V8 and SpiderMonkey.

For sloppy mode eval() in global scope, this change fixes a) binding creation for hoisted block-level
function declarations (Annex B) not to invoke [[Set]] on global object, and b) CanDeclareGlobalVar [3]
implementation to check only own properties, rather than performing [[Prototype]] lookup.

For global code, this change introduces CanDeclareGlobalVar checks that throw if the global object
is both non-extensible and doesn&apos;t have the given property, which is observable only in non-browser
environments where extension of global object can actually be prevented.

Performance analysis for declaration in global code:
  * `function` before: structure lookup, symbol table lookup [x2], add symbol table entry;
  * `function` after: structure lookup [x2], symbol table lookup [x2], add symbol table entry;
  * `var` before: structure lookup, symbol table lookup, symbol table lookup, add symbol table entry;
  * `var` after: structure lookup, symbol table lookup, add symbol table entry.

While this change adds extra structure lookup for `function`, it removes a few virtual calls, as well
as calling into JSLocalDOMWindow::getOwnPropertySlot().

As demonstrated by added tests, merging CanDeclareGlobalX with CreateGlobalXBinding is impossible since
no binding should be created, which is observable after catching an error, even if a single CanDeclareGlobalX
check fails. However, follow-up optimizations to reduce structure / symbol table lookups are possible.

All added / modified tests are were proven to align JSC with the spec, SpiderMonkey, and V8 (for the most part).

[1] <a href="https://tc39.es/ecma262/#sec-candeclareglobalfunction">https://tc39.es/ecma262/#sec-candeclareglobalfunction</a>
[2] <a href="https://tc39.es/ecma262/#sec-invariants-of-the-essential-internal-methods">https://tc39.es/ecma262/#sec-invariants-of-the-essential-internal-methods</a>
[3] <a href="https://tc39.es/ecma262/#sec-createglobalfunctionbinding">https://tc39.es/ecma262/#sec-createglobalfunctionbinding</a>
[4] <a href="https://tc39.es/ecma262/#sec-candeclareglobalvar">https://tc39.es/ecma262/#sec-candeclareglobalvar</a>

* JSTests/stress/can-declare-global-function-invoked-before-any-binding-is-created-eval.js: Added.
* JSTests/stress/can-declare-global-function-invoked-before-any-binding-is-created-global.js: Added.
* JSTests/stress/can-declare-global-function-invoked-before-any-func-decl-is-hoisted-eval.js: Added.
* JSTests/stress/can-declare-global-function-invoked-before-can-declare-var-eval.js: Added.
* JSTests/stress/can-declare-global-function-invoked-before-can-declare-var-global.js: Added.
* JSTests/stress/can-declare-global-function-invoked-var-by-the-same-name-eval.js: Added.
* JSTests/stress/can-declare-global-function-invoked-var-by-the-same-name-global.js: Added.
* JSTests/stress/can-declare-global-var-invoked-before-any-binding-is-created-eval.js: Added.
* JSTests/stress/can-declare-global-var-invoked-before-any-binding-is-created-global.js: Added.
* JSTests/stress/can-declare-global-var-invoked-before-any-func-decl-is-hoisted-eval.js: Added.
* JSTests/stress/can-declare-global-var-invoked-during-func-decl-hoisting-eval.js: Added.
* JSTests/stress/can-declare-global-var-invoked-from-jsonp-fast-path.js: Added.
* JSTests/stress/create-global-function-binding-updates-descriptor-eval.js: Added.
* JSTests/stress/create-global-function-binding-updates-descriptor-global.js: Added.
* JSTests/stress/eval-func-decl-block-does-not-invoke-setter.js: Added.
* JSTests/test262/expectations.yaml: Mark 21 tests as passing.
* LayoutTests/js/dom/var-declarations-shadowing-expected.txt:
* LayoutTests/js/dom/var-declarations-shadowing.html:
* LayoutTests/js/dom/function-declarations-shadowing-expected.txt: Added.
* LayoutTests/js/dom/function-declarations-shadowing.html: Added.
* Source/JavaScriptCore/bytecompiler/BytecodeGenerator.cpp:
(JSC::BytecodeGenerator::BytecodeGenerator):
* Source/JavaScriptCore/interpreter/Interpreter.cpp:
(JSC::Interpreter::executeProgram):
(JSC::Interpreter::executeEval):
* Source/JavaScriptCore/runtime/ExceptionHelpers.cpp:
(JSC::createErrorForInvalidGlobalFunctionDeclaration):
(JSC::createErrorForInvalidGlobalVarDeclaration):
* Source/JavaScriptCore/runtime/ExceptionHelpers.h:
* Source/JavaScriptCore/runtime/JSGlobalObject.cpp:
(JSC::JSGlobalObject::canDeclareGlobalFunction):
(JSC::JSGlobalObject::createGlobalFunctionBinding):
(JSC::JSGlobalObject::addSymbolTableEntry):
(JSC::JSGlobalObject::addGlobalVar): Deleted.
(JSC::JSGlobalObject::addFunction): Deleted.
* Source/JavaScriptCore/runtime/JSGlobalObject.h:
* Source/JavaScriptCore/runtime/JSGlobalObjectInlines.h:
(JSC::JSGlobalObject::canDeclareGlobalVar):
(JSC::JSGlobalObject::createGlobalVarBinding):
(JSC::JSGlobalObject::addVar): Deleted.
* Source/JavaScriptCore/runtime/ProgramExecutable.cpp:
(JSC::ProgramExecutable::initializeGlobalProperties):

Canonical link: <a href="https://commits.webkit.org/267655@main">https://commits.webkit.org/267655@main</a>
</pre>
<!--EWS-Status-Bubble-Start-->
https://github.com/WebKit/WebKit/commit/c6cfb13b7f7e1c67e508a19cfb3ce940ae5f9325

| Misc | iOS, tvOS & watchOS  | macOS  | Linux |  Windows |
| ----- | ---------------------- | ------- |  ----- |  --------- |
| [✅ 🧪 style](https://ews-build.webkit.org/#/builders/38/builds/17294 "Passed style check") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/26/builds/17619 "Built successfully") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/14/builds/18123 "Built successfully") | [✅ 🛠 wpe](https://ews-build.webkit.org/#/builders/5/builds/19082 "Built successfully") | [✅ 🛠 wincairo](https://ews-build.webkit.org/#/builders/32/builds/16182 "Built successfully") 
| | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/23/builds/20894 "Built successfully") | [✅ 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/16/builds/17762 "Built successfully") | [  ~~🧪 wpe-wk2~~](https://ews-build.webkit.org/#/builders/34/builds/18373 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/11/builds/17498 "Passed tests") | [✅ 🧪 ios-wk2](https://ews-build.webkit.org/#/builders/15/builds/17838 "Passed tests") | [✅ 🧪 api-mac](https://ews-build.webkit.org/#/builders/18/builds/15036 "Passed tests") | [✅ 🛠 gtk](https://ews-build.webkit.org/#/builders/2/builds/19899 "Built successfully") | 
| | [✅ 🧪 ios-wk2-wpt](https://ews-build.webkit.org/#/builders/39/builds/15078 "Passed tests") | [  ~~🧪 mac-wk1~~](https://ews-build.webkit.org/#/builders/10/builds/15726 "The change is no longer eligible for processing. Commit was outdated when EWS attempted to process it.") | [  ~~🧪 gtk-wk2~~](https://ews-build.webkit.org/#/builders/1/builds/22400 "The change is no longer eligible for processing. Commit was outdated when EWS attempted to process it.") | 
| [✅ 🛠 🧪 jsc](https://ews-build.webkit.org/#/builders/20/builds/14944 "Built successfully and passed tests") | [✅ 🧪 api-ios](https://ews-build.webkit.org/#/builders/13/builds/16077 "Passed tests") | [  ~~🧪 mac-wk2~~](https://ews-build.webkit.org/#/builders/36/builds/15894 "The change is no longer eligible for processing. Commit was outdated when EWS attempted to process it.") | [✅ 🧪 api-gtk](https://ews-build.webkit.org/#/builders/21/builds/20224 "Passed tests") | 
| [✅ 🛠 🧪 jsc-arm64](https://ews-build.webkit.org/#/builders/12/builds/16514 "Built successfully and passed tests") | [✅ 🛠 tv](https://ews-build.webkit.org/#/builders/7/builds/16477 "Built successfully") | [  ~~🧪 mac-AS-debug-wk2~~](https://ews-build.webkit.org/#/builders/17/builds/13986 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [✅ 🛠 jsc-armv7](https://ews-build.webkit.org/#/builders/35/builds/18857 "Built successfully") | 
| | [✅ 🛠 tv-sim](https://ews-build.webkit.org/#/builders/30/builds/15624 "Built successfully") | [✅ 🧪 mac-wk2-stress](https://ews-build.webkit.org/#/builders/8/builds/15566 "Passed tests") | [✅ 🧪 jsc-armv7-tests](https://ews-build.webkit.org/#/builders/25/builds/4376 "Passed tests") | 
| | [✅ 🛠 watch](https://ews-build.webkit.org/#/builders/4/builds/19994 "Built successfully") | | [✅ 🛠 jsc-mips](https://ews-build.webkit.org/#/builders/24/builds/20083 "Built successfully") | 
| [✅ 🛠 🧪 unsafe-merge](https://ews-build.webkit.org/#/builders/22/builds/2113 "Built successfully and passed tests") | [✅ 🛠 watch-sim](https://ews-build.webkit.org/#/builders/31/builds/16312 "Built successfully") | | [❌ 🧪 jsc-mips-tests](https://ews-build.webkit.org/#/builders/3/builds/4231 "Found 1 jsc stress test failure: microbenchmarks/array-from-object-func.js.lockdown") | 
<!--EWS-Status-Bubble-End-->

Constellation reviewed (approved):
r=me if EWS gets green.

Commit:
[JSC] Implement CanDeclareGlobalFunction abstract operation and friends
https://bugs.webkit.org/show_bug.cgi?id=260487
<rdar://problem/114215396>

Reviewed by Yusuke Suzuki.

For both global code and sloppy mode eval() in global scope, this change adds CanDeclareGlobalFunction [1]
that checks if a function can be declared while conforming to invariants of essential internal methods [2],
and also introduces CreateGlobalFunctionBinding [3] that creates a binding with best possible descriptor,
attempting to make the property both writable and enumerable.

For global code, the presence of these helpers is observable when attempting to shadow a non-configurable
global propeperty, and in browser environments with multiple <script> elements: if first script defines
non-configurable property X, the second one containing `function X() {}` should throw a TypeError.
In Bun, they can be observed via vm.runInThisContext() API.

Just like before the change, global code functions are defined on the symbol table unless there is a
non-configurable structure property, ensuring the best possible performance. CanDeclareGlobalFunction
and CreateGlobalFunctionBinding substitute the usage of non-idiomatic VM::DeletePropertyModeScope hack,
which used to ensure there is no property by the same name both on the structure and on the symbol table.
This change makes its removal one step closer.

Functions declared via sloppy mode eval() in global scope are unconditionally defined on the structure
because JSC relies on symbol table entries being immutable. This also preserves a bit weird global object
property enumeration order that is different from both V8 and SpiderMonkey.

For sloppy mode eval() in global scope, this change fixes a) binding creation for hoisted block-level
function declarations (Annex B) not to invoke [[Set]] on global object, and b) CanDeclareGlobalVar [3]
implementation to check only own properties, rather than performing [[Prototype]] lookup.

For global code, this change introduces CanDeclareGlobalVar checks that throw if the global object
is both non-extensible and doesn't have the given property, which is observable only in non-browser
environments where extension of global object can actually be prevented.

Performance analysis for declaration in global code:
  * `function` before: structure lookup, symbol table lookup [x2], add symbol table entry;
  * `function` after: structure lookup [x2], symbol table lookup [x2], add symbol table entry;
  * `var` before: structure lookup, symbol table lookup, symbol table lookup, add symbol table entry;
  * `var` after: structure lookup, symbol table lookup, add symbol table entry.

While this change adds extra structure lookup for `function`, it removes a few virtual calls, as well
as calling into JSLocalDOMWindow::getOwnPropertySlot().

As demonstrated by added tests, merging CanDeclareGlobalX with CreateGlobalXBinding is impossible since
no binding should be created, which is observable after catching an error, even if a single CanDeclareGlobalX
check fails. However, follow-up optimizations to reduce structure / symbol table lookups are possible.

All added / modified tests are were proven to align JSC with the spec, SpiderMonkey, and V8 (for the most part).

[1] https://tc39.es/ecma262/#sec-candeclareglobalfunction
[2] https://tc39.es/ecma262/#sec-invariants-of-the-essential-internal-methods
[3] https://tc39.es/ecma262/#sec-createglobalfunctionbinding
[4] https://tc39.es/ecma262/#sec-candeclareglobalvar

* JSTests/stress/can-declare-global-function-invoked-before-any-binding-is-created-eval.js: Added.
* JSTests/stress/can-declare-global-function-invoked-before-any-binding-is-created-global.js: Added.
* JSTests/stress/can-declare-global-function-invoked-before-any-func-decl-is-hoisted-eval.js: Added.
* JSTests/stress/can-declare-global-function-invoked-before-can-declare-var-eval.js: Added.
* JSTests/stress/can-declare-global-function-invoked-before-can-declare-var-global.js: Added.
* JSTests/stress/can-declare-global-function-invoked-var-by-the-same-name-eval.js: Added.
* JSTests/stress/can-declare-global-function-invoked-var-by-the-same-name-global.js: Added.
* JSTests/stress/can-declare-global-var-invoked-before-any-binding-is-created-eval.js: Added.
* JSTests/stress/can-declare-global-var-invoked-before-any-binding-is-created-global.js: Added.
* JSTests/stress/can-declare-global-var-invoked-before-any-func-decl-is-hoisted-eval.js: Added.
* JSTests/stress/can-declare-global-var-invoked-during-func-decl-hoisting-eval.js: Added.
* JSTests/stress/can-declare-global-var-invoked-from-jsonp-fast-path.js: Added.
* JSTests/stress/create-global-function-binding-updates-descriptor-eval.js: Added.
* JSTests/stress/create-global-function-binding-updates-descriptor-global.js: Added.
* JSTests/stress/eval-func-decl-block-does-not-invoke-setter.js: Added.
* JSTests/test262/expectations.yaml: Mark 21 tests as passing.
* LayoutTests/js/dom/var-declarations-shadowing-expected.txt:
* LayoutTests/js/dom/var-declarations-shadowing.html:
* LayoutTests/js/dom/function-declarations-shadowing-expected.txt: Added.
* LayoutTests/js/dom/function-declarations-shadowing.html: Added.
* Source/JavaScriptCore/bytecompiler/BytecodeGenerator.cpp:
(JSC::BytecodeGenerator::BytecodeGenerator):
* Source/JavaScriptCore/interpreter/Interpreter.cpp:
(JSC::Interpreter::executeProgram):
(JSC::Interpreter::executeEval):
* Source/JavaScriptCore/runtime/ExceptionHelpers.cpp:
(JSC::createErrorForInvalidGlobalFunctionDeclaration):
(JSC::createErrorForInvalidGlobalVarDeclaration):
* Source/JavaScriptCore/runtime/ExceptionHelpers.h:
* Source/JavaScriptCore/runtime/JSGlobalObject.cpp:
(JSC::JSGlobalObject::canDeclareGlobalFunction):
(JSC::JSGlobalObject::createGlobalFunctionBinding):
(JSC::JSGlobalObject::addSymbolTableEntry):
(JSC::JSGlobalObject::addGlobalVar): Deleted.
(JSC::JSGlobalObject::addFunction): Deleted.
* Source/JavaScriptCore/runtime/JSGlobalObject.h:
* Source/JavaScriptCore/runtime/JSGlobalObjectInlines.h:
(JSC::JSGlobalObject::canDeclareGlobalVar):
(JSC::JSGlobalObject::createGlobalVarBinding):
(JSC::JSGlobalObject::addVar): Deleted.
* Source/JavaScriptCore/runtime/ProgramExecutable.cpp:
(JSC::ProgramExecutable::initializeGlobalProperties):

Canonical link: https://commits.webkit.org/267655@main
